x/merkledb/metrics.go

```diff
@@ -11,38 +11,176 @@ import (
+		lookupResult: hitResult,
+	}
+	viewChangesValueMissLabels = prometheus.Labels{
+		lookupType:   viewChangesValueType,
+		lookupResult: missResult,
+	}
+	viewChangesNodeHitLabels = prometheus.Labels{
+		lookupType:   viewChangesNodeType,
+		lookupResult: hitResult,
+	}
+	viewChangesNodeMissLabels = prometheus.Labels{
+		lookupType:   viewChangesNodeType,
+		lookupResult: missResult,
+	}
 )
 
-type merkleMetrics interface {
+type metrics interface {
+	HashCalculated()
 	DatabaseNodeRead()
 	DatabaseNodeWrite()
-	HashCalculated()
 	ValueNodeCacheHit()
 	ValueNodeCacheMiss()
 	IntermediateNodeCacheHit()
 	IntermediateNodeCacheMiss()
-	ViewChangesNodeHit()
-	ViewChangesNodeMiss()
 	ViewChangesValueHit()
 	ViewChangesValueMiss()
+	ViewChangesNodeHit()
+	ViewChangesNodeMiss()
+}
+
+type prometheusMetrics struct {
+	hashes prometheus.Counter
+	io     *prometheus.CounterVec
+	lookup *prometheus.CounterVec
+}
+
+func newMetrics(namespace string, reg prometheus.Registerer) (metrics, error) {
+	// TODO: Should we instead return an error if reg is nil?
+	if reg == nil {
+		return &mockMetrics{}, nil
+	}
+	m := prometheusMetrics{
+		hashes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "hashes",
+			Help:      "cumulative number of nodes hashed",
+		}),
+		io: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "io",
+			Help:      "cumulative number of operations performed to the db",
+		}, ioLabels),
+		lookup: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "lookup",
+			Help:      "cumulative number of in-memory lookups performed",
+		}, lookupLabels),
+	}
+	err := utils.Err(
+		reg.Register(m.hashes),
+		reg.Register(m.io),
+		reg.Register(m.lookup),
+	)
+	return &m, err
+}
+
+func (m *prometheusMetrics) HashCalculated() {
+	m.hashes.Inc()
+}
+
+func (m *prometheusMetrics) DatabaseNodeRead() {
+	m.io.With(ioReadLabels).Inc()
+}
+
+func (m *prometheusMetrics) DatabaseNodeWrite() {
+	m.io.With(ioWriteLabels).Inc()
+}
+
+func (m *prometheusMetrics) ValueNodeCacheHit() {
+	m.lookup.With(valueNodeCacheHitLabels).Inc()
+}
+
+func (m *prometheusMetrics) ValueNodeCacheMiss() {
+	m.lookup.With(valueNodeCacheMissLabels).Inc()
+}
+
+func (m *prometheusMetrics) IntermediateNodeCacheHit() {
+	m.lookup.With(intermediateNodeCacheHitLabels).Inc()
+}
+
+func (m *prometheusMetrics) IntermediateNodeCacheMiss() {
+	m.lookup.With(intermediateNodeCacheMissLabels).Inc()
+}
+
+func (m *prometheusMetrics) ViewChangesValueHit() {
+	m.lookup.With(viewChangesValueHitLabels).Inc()
+}
+
+func (m *prometheusMetrics) ViewChangesValueMiss() {
+	m.lookup.With(viewChangesValueMissLabels).Inc()
+}
+
+func (m *prometheusMetrics) ViewChangesNodeHit() {
+	m.lookup.With(viewChangesNodeHitLabels).Inc()
+}
+
+func (m *prometheusMetrics) ViewChangesNodeMiss() {
+	m.lookup.With(viewChangesNodeMissLabels).Inc()
 }
 
 type mockMetrics struct {
```

Comment:
I felt like this should be defined below the actual implementation... Because when people normally look at a file they start at the top.